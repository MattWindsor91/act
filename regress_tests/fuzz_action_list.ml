(* The Automagic Compiler Tormentor

   Copyright (c) 2018--2019 Matt Windsor and contributors

   ACT itself is licensed under the MIT License. See the LICENSE file in the
   project root for more information.

   ACT is based in part on code from the Herdtools7 project
   (https://github.com/herd/herdtools7) : see the LICENSE.herd file in the
   project root for more information. *)

(** Lists the available fuzz actions, their default weights, and their names
    and READMEs. Intended to track unwanted changes to the action pool. *)

open Core

(* TODO(@MattWindsor91): merge with `act-fuzz list_actions`? *)

let run (_test_dir : Fpath.t) : unit Or_error.t =
  Or_error.(
    Act_fuzz_run.Config.(make () |> summarise)
    >>| Fmt.pr "@[<v>%a@]@." Act_fuzz.Action.Summary.pp_map)

let command : Command.t =
  Common.make_regress_command ~summary:"runs fuzz action list regressions"
    run

let () = Command.run command
