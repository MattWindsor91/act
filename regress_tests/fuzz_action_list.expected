flow.dead.early-out:
  Weight: 15x
  Summary:
    Inserts a valid 'early-out' statement (break or return) into a random
    dead-code location.
flow.if.duplicate:
  Weight: 15x
  Summary:
    Removes a sublist of statements from the program, replacing them with an
    `if` statement containing some transformation of the removed
    statements.
    
    This version of the action generates an arbitrary condition, and
    initialises both branches of the `if` statement with the original
    statements.
flow.if.tautology:
  Weight: 25x
  Summary:
    Removes a sublist of statements from the program, replacing them with an
    `if` statement containing some transformation of the removed
    statements.
    
    This version of the action generates an always-true condition, puts the
    original statements in the true block, and marks the false block as
    dead-code.
flow.invert-if:
  Weight: 10x
  Summary: Flips the conditional and branches of an if statement.
flow.loop.surround:
  Weight: 20x
  Summary:
    Removes a sublist of statements from the program, replacing them with a
    `do... while` statement containing some transformation of the removed
    statements.
    
    The condition of the `do... while` loop is statically guaranteed to be
    false.
mem.fence:
  Weight: 15x
  Summary: Inserts a randomly generated memory fence into the test.
program.label:
  Weight: 10x
  Summary: Inserts a new, random label into the program.
program.make.empty:
  Weight: 10x
  Summary:
    Generates a new, empty program at one end of the program list. This
    action isn't very useful on its own, but works well in combination with
    other actions that construct statements and control flows.
store.make.int.dead:
  Weight: 20x
  Summary:
    Generates a store operation on a randomly selected fuzzer-generated
    global variable.
    
    If 'store.forbid-already-written' is true, this action will only store to
    variables that haven't previously been selected for store actions. This
    makes calculating candidate executions easier, but limits the degree of
    entropy somewhat. (Note that if the value is stochastic, the action will
    only fire if such variables exist, but may or may not proceed to select a
    previously-stored variable. This is a limitation of the flag system.)
    
    This variant can target any source and destination, but only inserts into
    dead code. As it only targets dead code, it does not add dependences or
    erase known-values.
    
    This operation generates 'atomic_int's.
store.make.int.normal:
  Weight: 30x
  Summary:
    Generates a store operation on a randomly selected fuzzer-generated
    global variable.
    
    If 'store.forbid-already-written' is true, this action will only store to
    variables that haven't previously been selected for store actions. This
    makes calculating candidate executions easier, but limits the degree of
    entropy somewhat. (Note that if the value is stochastic, the action will
    only fire if such variables exist, but may or may not proceed to select a
    previously-stored variable. This is a limitation of the flag system.)
    
    This variant can insert anywhere and target any source and
    destination.
    
    This operation generates 'atomic_int's.
store.make.int.redundant:
  Weight: 15x
  Summary:
    Generates a store operation on a randomly selected fuzzer-generated
    global variable.
    
    If 'store.forbid-already-written' is true, this action will only store to
    variables that haven't previously been selected for store actions. This
    makes calculating candidate executions easier, but limits the degree of
    entropy somewhat. (Note that if the value is stochastic, the action will
    only fire if such variables exist, but may or may not proceed to select a
    previously-stored variable. This is a limitation of the flag system.)
    
    This variant can insert anywhere, but only stores the known value of a
    destination back to itself.
    
    This operation generates 'atomic_int's.
var.make.global:
  Weight: 20x
  Summary:
    Generates a new global variable, with a random name, initial value, and
    primitive type.
